runtime/lua/vim/lsp/_snippet.lua

```diff
@@ -0,0 +1,421 @@
+    P.seq(S.plus, P.take_until({ '}' }, { '\\' })),
+    P.seq(S.minus, P.take_until({ '}' }, { '\\' }))
+  ),  S.close), function(values)
+    return {
+      type = NodeType.FORMAT,
+      capture_index = values[3],
+      if_text = values[5][2].esc,
+      else_text = (values[5][4] or {}).esc,
+    }
+  end)
+)
+
+S.transform = P.map(P.seq(
+  S.slash,
+  P.take_until({ '/' }, { '\\' }),
+  S.slash,
+  P.many(P.any(S.format, S.text({ '$', '/' }, { '\\' }))),
+  S.slash,
+  P.pattern('[ig]+')
+), function(values)
+  return {
+    type = NodeType.TRANSFORM,
+    pattern = values[2].raw,
+    format = values[4],
+    option = values[6],
+  }
+end)
+
+S.tabstop = P.any(
+  P.map(P.seq(S.dollar, S.int), function(values)
+    return {
+      type = NodeType.TABSTOP,
+      tabstop = values[2],
+    }
+  end),
+  P.map(P.seq(S.dollar, S.open, S.int, S.close), function(values)
+    return {
+      type = NodeType.TABSTOP,
+      tabstop = values[3],
+    }
+  end),
+  P.map(P.seq(S.dollar, S.open, S.int, S.transform, S.close), function(values)
+    return {
+      type = NodeType.TABSTOP,
+      tabstop = values[3],
+      transform = values[4]
+    }
+  end)
+)
+
+S.placeholder = P.any(
+  P.map(P.seq(S.dollar, S.open, S.int, S.colon, P.many(P.any(S.toplevel, S.text({ '$', '}' }, { '\\' }))), S.close), function(values)
+    return {
+      type = NodeType.PLACEHOLDER,
+      tabstop = values[3],
+      children = values[5],
+    }
+  end)
+)
+
+S.choice = P.map(P.seq(
+  S.dollar,
+  S.open,
+  S.int,
+  S.pipe,
+  P.many(
+    P.map(P.seq(S.text({ ',', '|' }), P.opt(S.comma)), function(values)
+      return values[1].esc
+    end)
+  ),
+  S.pipe,
+  S.close
+), function(values)
+  return {
+    type = NodeType.CHOICE,
+    tabstop = values[3],
+    items = values[5],
+  }
+end)
+
+S.variable = P.any(
+  P.map(P.seq(S.dollar, S.var), function(values)
+    return {
+      type = NodeType.VARIABLE,
+      name = values[2],
+    }
+  end),
+  P.map(P.seq(S.dollar, S.open, S.var, S.close), function(values)
+    return {
+      type = NodeType.VARIABLE,
+      name = values[3],
+    }
+  end),
+  P.map(P.seq(S.dollar, S.open, S.var, S.transform, S.close), function(values)
+    return {
+      type = NodeType.VARIABLE,
+      name = values[3],
+      transform = values[4],
+    }
+  end),
+  P.map(P.seq(S.dollar, S.open, S.var, S.colon, P.many(P.any(S.toplevel, S.text({ '$', '}' }, { '\\' }))), S.close), function(values)
+    return {
+      type = NodeType.VARIABLE,
+      name = values[3],
+      children = values[5],
+    }
+  end)
+)
+
+S.snippet = P.map(P.many(P.any(S.toplevel, S.text({ '$' }, { '}', '\\' }))), function(values)
+  return {
+    type = NodeType.SNIPPET,
+    children = values,
+  }
+end)
+
+local M = {}
+
+---@types table<string, number>
+---The snippet node type enum
+M.NodeType = NodeType
+
+---Parse snippet string and returns the AST
+---@param input string
+---@return table
+M.parse = function(input)
```

Comment:
is there a reason not to use the `function M.parse(input)` form? that is the convention in most of the codebase.

Reply:
It has no reason. I will change it!

Reply:
I think this is the other holdup, after this and the other comment is addressed regarding adding a fmt convention I am good to merge :)